dimod/cyvariables.pyx

```diff
@@ -0,0 +1,263 @@
+            else:
+                raise ValueError('{!r} is already a variable'.format(v))
+
+        idx = self._stop
+
+        if idx != v:
+            self._label_to_index[v] = idx
+            self._index_to_label[idx] = v
+
+        self._stop += 1
+        return v
+
+    cpdef bint _is_range(self):
+        """Return whether the Variables are current labelled [0, n)."""
+        return not PyDict_Size(self._label_to_index)
+
+    cpdef object _extend(self, object iterable, bint permissive=False):
+        """Add new variables.
+
+        Args:
+            iterable (iterable[hashable], optional):
+                An iterable of hashable objects.
+
+            permissive (bool, optional, default=False):
+                If `False`, appending a variable that already exists will raise
+                a `ValueError`. If `True`, appending a variable that already
+                exists will not change the container.
+
+        Raises:
+            ValueError: If a variable is present and `permissive` is
+            False.
+
+        This method is semi-public. it is intended to be used by
+        classes that have :class:`.Variables` as an attribute, not by the
+        the user.
+        """
+        # todo: performance improvements for range etc
+        for v in iterable:
+            self._append(v, permissive=permissive)
+
+    def _pop(self):
+        """Remove the last variable.
+
+        This method is semi-public. it is intended to be used by
+        classes that have :class:`.Variables` as an attribute, not by the
+        the user.
+        """
+        if not self:
+            raise IndexError("Cannot pop when Variables is empty")
+
+        self._stop = idx = self._stop - 1
+
+        label = self._index_to_label.pop(idx, idx)
+        self._label_to_index.pop(label, None)
+        return label
+
+    def _relabel(self, mapping):
+        """Relabel the variables in-place.
+
+        Args:
+            mapping (dict):
+                Mapping from current variable labels to new, as a dict. If
+                an incomplete mapping is specified, unmapped variables keep
+                their current labels.
+
+        This method is semi-public. it is intended to be used by
+        classes that have :class:`.Variables` as an attribute, not by the
+        the user.
+        """
+        for submap in iter_safe_relabels(mapping, self):
+            for old, new in submap.items():
+                if old == new:
+                    continue
+
+                idx = self._label_to_index.pop(old, old)
+
+                if new != idx:
+                    self._label_to_index[new] = idx
+                    self._index_to_label[idx] = new  # overwrites old idx
+                else:
+                    self._index_to_label.pop(idx, None)
+
+    def _relabel_as_integers(self):
+        """Relabel the variables as integers in-place.
+
+        Returns:
+            dict: The mapping that would restore the original labels.
```

Comment:
What does "original" mean here? Are the `self._label_to_index` and `self._index_to_label` keeping a history of label indices?